docs/sphinx/user_guide/feature/array.rst

```diff
@@ -0,0 +1,170 @@
+
+    camp::array<double, N> arr;
+
+It's that simple!
+
+.. important:: 
+  While the Camp Array tries to emulate the STL array as closely as it can, there are some key differences 
+  that users should be aware of: 
+  
+  * no ``__device__ "at"`` method. The ``at`` method has the potential to throw exceptions, which is not GPU friendly.
+  * Calling ``front()`` or ``back()`` is a compile time error when ``size() == 0``, 
+    as opposed to undefined  behaviour set by the C++ standard.
+  * camp arrays do not have reverse iterators implemented. 
+  * camp arrays do not have the ``swap`` method implemented.   
+
+Constructors
+^^^^^^^^^^^^
+
+Camp arrays, like their STL counterparts, require the size and the type to be known at compile time. The array constructor is as follows:
+  
+.. code-block:: cpp
+
+    camp::array<class T, std::size_t N>;
+
+where T and N are provided by the user. In our previous example, we created an array of 100 ``doubles``:
+  
+.. code-block:: cpp
+
+    constexpr size_t N = 100; 
+
+    camp::array<double, N> arr;
+
+We can also initialize our array at compile time, if we know the values:
+
+.. code-block:: cpp
+  
+    camp::array<int, 3> arr = {1, 2, 3};
+
+Element Access
+^^^^^^^^^^^^^^
+
+To access an element of the array, we can either use the ``operator[]``, or the ``at`` methods. The only real difference between these 
+two methods is that the ``at`` method performs bounds checking on the input, and will throw an exception if the index requested is out of 
+the bounds of the array.
+
+.. code-block:: cpp
+
+    camp::array<int, 3> arr  = {1, 2, 3};
+    std::cout << arr[1] << std::endl; // prints 2
+    std::cout << arr.at(2) << std::endl; // prints 3
+    std::cout << arr.at(4) << std::endl; // throws std::out_of_range
+    std::cout << arr[4] << std::endl; // undefined behaviour; accesses illegal memory
+
+The ``front()`` and ``back()`` methods can be used to obtain a reference (const or non-const) to the first and last element in the array, respectively:
+
+.. code-block:: cpp
+
+    camp::array<int, 3> arr  = {1, 2, 3};
+    std::cout << "front is " << arr.front() << ", back is " << arr.back() <<std::endl;
+    // output: front is 1, back is 3
+
+    arr.front() = 4;
+    // arr is now {4, 2, 3}
+
+A pointer to the underlying data can be obtained using the ``data()`` method. Camp::array implements ``begin(), end(), cbegin(),`` and ``cend()`` iterator functions, which allow it to be used interchangeably in many c++ algorithms in the standard library, and beyond:
+
+.. code-block:: cpp
+  
+    camp::array<int, 3> arr  = {1, 2, 3};
+    for (const auto elem : arr) { 
+      std::cout << elem << "\n";
+    }
+
+Camp provides multiple ``get`` methods that can be used for constexpr element access and moving:
+
+.. code-block:: cpp
+
+    camp::array<int, 3> arr  = {1, 2, 3};
+    return camp::get<1>(arr); // returns 2
+
+``get`` can return const and non-const lvalue references, and rvalue references. 
```

Comment:
Simple examples of these would be good to add.

Reply:
(Since the above comment had a committed suggestion, it comes up as resolved - I am copy-pasting the comment part here too to keep track of it...)
....
Line 130: Also, an example of using the data() method similar to the others where you have to derefence the pointer (via [] or *) would be good to add.